### PR TITLE
idle: move idleness manager to separate package and ~13s of tests into it

### PR DESCRIPTION
**Pull request by dfawley**
Also: pull out some shared testing code to make it easier to use in other packages than `test/`.

Before:

```
$ time go test -count=1 ./...
<snip>
real	0m47.221s
user	1m31.381s
sys	0m30.061s
```

After:

```
$ time go test -count=1 ./...
<snip>
real	0m38.994s
user	1m30.279s
sys	0m29.964s
```

For reference:

```
$ go test -count=1 ./internal/idle
ok  	google.golang.org/grpc/internal/idle	12.739s
```

RELEASE NOTES: none

### CODE REVIEW

File: internal/idle/idle.go [hunk 1]
```diff
@@ -16,41 +16,45 @@
...
+	"google.golang.org/grpc/grpclog"
 )
 
 // For overriding in unit tests.
 var timeAfterFunc = func(d time.Duration, f func()) *time.Timer {
 	return time.AfterFunc(d, f)
 }
 
-// idlenessEnforcer is the functionality provided by grpc.ClientConn to enter
+// IdlenessEnforcer is the functionality provided by grpc.ClientConn to enter
```
Comment (easwars) [inline internal/idle/idle.go:38]: Should all of these names not include the `Idleness` prefix?

I'm ok with names being `idle.Enforcer`, `idle.Manager`, and `noopManager`, `managerImpl`, `idle.ManagerOptions` and `idle.NewManager()`.

If you feel that is not descriptive enough, we can rename the package `idleness` and then these names would be `idleness.Enforcer` and so on.

Wdyt?

File: internal/idle/idle_test.go [hunk 2]
```diff
@@ -245,29 +257,29 @@ func (s) TestIdlenessManager_Enabled_ExitIdleOnRPC(t *testing.T) {
 	overrideNewTimer(t)
 
 	enforcer := newTestIdlenessEnforcer()
-	mgr := newIdlenessManager(enforcer, time.Duration(defaultTestIdleTimeout))
-	defer mgr.close()
+	mgr := NewIdlenessManager(IdlenessManagerOptions{Enforcer: enforcer, Timeout: time.Duration(defaultTestIdleTimeout), Logger: grpclog.Component("test")})
```
Comment (easwars) [inline internal/idle/idle_test.go:260]: That is interesting. I have usually passed a `nil` logger in tests. Does this result in logs being any different in the tests?

File: internal/testutils/state.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+func AwaitState(ctx context.Context, t *testing.T, sc StateChanger, stateWant connectivity.State) {
+	t.Helper()
+	for state := sc.GetState(); state != stateWant; state = sc.GetState() {
+		if !sc.WaitForStateChange(ctx, state) {
+			t.Fatalf("timed out waiting for state change.  got %v; want %v", state, stateWant)
+		}
+	}
+}
+
+// AwaitNotState waits for sc to leave stateWant or fatal errors if it doesn't
```
Comment (easwars) [inline internal/testutils/state.go:67]: s/stateWant/stateDoNotWant

File: internal/testutils/state.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	}
+}
+
+// AwaitState waits for sc to enter stateWant or fatal errors if it doesn't
+// happen before ctx expires.
+func AwaitState(ctx context.Context, t *testing.T, sc StateChanger, stateWant connectivity.State) {
+	t.Helper()
+	for state := sc.GetState(); state != stateWant; state = sc.GetState() {
+		if !sc.WaitForStateChange(ctx, state) {
+			t.Fatalf("timed out waiting for state change.  got %v; want %v", state, stateWant)
```
Comment (easwars) [inline internal/testutils/state.go:62]: Here and in the next t.Fatalf(), s/timed/Timed/

File: internal/testutils/state.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		case connectivity.Shutdown:
+			return
+		}
+		if !sc.WaitForStateChange(ctx, state) {
+			return
+		}
+	}
+}
+
+// AwaitState waits for sc to enter stateWant or fatal errors if it doesn't
```
Comment (easwars) [inline internal/testutils/state.go:56]: We seem to be in direct violation of this guideline: go/go-style/decisions#assert.

File: internal/testutils/wrappers.go [hunk 6]
```diff
@@ -72,3 +72,13 @@ func NewListenerWrapper(t *testing.T, lis net.Listener) *ListenerWrapper {
 		NewConnCh: NewChannel(),
 	}
 }
+
+// ErrCloseWrapper wraps closer with a function that does not return an error,
+// but calls t.Error if it does, instead.
+func ErrCloseWrapper(t *testing.T, closer func() error) func() {
```
Comment (easwars) [inline internal/testutils/wrappers.go:78]: Same here. If the `closer` doesn't return an informative error, we don't have enough contextual information here to push an informative error message to the test output.

**Comment by dfawley** [comment]
Sorry, my `push` failed because I made a commit on my laptop that I hadn't pulled onto my workstation.  This is ready to review now.

**Comment by dfawley** [comment]
Wait now that I did that, my tests are failing, sorry!

File: clientconn.go [hunk 7]
```diff
@@ -266,7 +267,7 @@ func DialContext(ctx context.Context, target string, opts ...DialOption) (conn *
 	// Configure idleness support with configured idle timeout or default idle
 	// timeout duration. Idleness can be explicitly disabled by the user, by
 	// setting the dial option to 0.
-	cc.idlenessMgr = newIdlenessManager(cc, cc.dopts.idleTimeout)
+	cc.idlenessMgr = idle.NewManager(idle.ManagerOptions{Enforcer: (*idler)(cc), Timeout: cc.dopts.idleTimeout, Logger: logger})
```
Comment (easwars) [inline clientconn.go:270]: Just curious why you find this single line literal struct initialization more readable than the more common multiline one.

File: internal/idle/idle.go [hunk 8]
```diff
@@ -189,99 +203,99 @@ func (i *idlenessManagerImpl) tryEnterIdleMode() bool {
...
 }
 
-// exitIdleMode instructs the channel to exit idle mode.
+// ExitIdleMode instructs the channel to exit idle mode.
 //
 // Holds idleMu which ensures mutual exclusion with tryEnterIdleMode.
-func (i *idlenessManagerImpl) exitIdleMode() error {
-	i.idleMu.Lock()
-	defer i.idleMu.Unlock()
+func (m *manager) ExitIdleMode() error {
```
Comment (easwars) [inline internal/idle/idle.go:244]: We don't need to export this method, right?

File: internal/idle/idle_test.go [hunk 9]
```diff
@@ -25,34 +25,46 @@ import (
...
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+type testEnforcer struct {
+	ExitIdleCh  chan struct{}
+	EnterIdleCh chan struct{}
```
Comment (easwars) [inline internal/idle/idle_test.go:49]: Do these channels have to be exported?

File: internal/idle/idle.go (hunk 8 above)
Comment (dfawley) [reply internal/idle/idle.go:244]: Good call; fixed.

File: internal/idle/idle_test.go (hunk 9 above)
Comment (dfawley) [reply internal/idle/idle_test.go:49]: Whoops.  Changed back.

File: clientconn.go (hunk 7 above)
Comment (dfawley) [reply clientconn.go:270]: IDK personally I find the parameters passed in here are really obvious and uninteresting, and I'd rather have one operation per line more (for denser code, which is easier to scroll through) than I'd like to be able to more spaciously see all the details of the values.  If I needed to do tricky math or call a function to set them, I'd probably make it multi-line.

Same reason I dislike most `var` blocks, particularly in functions, e.g. 

```go
var (
	i int
	p peer.Peer
)
```

Those variables aren't even related; why group them when it takes even more lines than declaring them separately?

**Review by easwars — APPROVED** [review]

**Comment by dfawley** [comment]
The changes to testing methodology of channelz turned up a latent race in the deletion code:
```
sync.runtime_SemacquireMutex(0x7f85605b88e0?, 0xc0?, 0x4167b4?)
	/opt/hostedtoolcache/go/1.20.7/x64/src/runtime/sema.go:77 +0x26
sync.(*Mutex).lockSlow(0xc0002ffe28)
	/opt/hostedtoolcache/go/1.20.7/x64/src/sync/mutex.go:171 +0x165
sync.(*Mutex).Lock(...)
	/opt/hostedtoolcache/go/1.20.7/x64/src/sync/mutex.go:90
google.golang.org/grpc/internal/channelz.(*channelTrace).clear(0xc0002ffe00)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:659 +0x45
google.golang.org/grpc/internal/channelz.(*channel).deleteSelfIfReady(0xc0009f1d40)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:288 +0x1d3
google.golang.org/grpc/internal/channelz.(*channel).deleteChild(0xc0009f1d40, 0xc0001d3470?)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:224 +0x58
google.golang.org/grpc/internal/channelz.(*subChannel).deleteSelfFromTree(0xc000df6480)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:357 +0x15d
google.golang.org/grpc/internal/channelz.(*subChannel).deleteSelfIfReady(0xc000df6480)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:389 +0x25
google.golang.org/grpc/internal/channelz.(*channelMap).decrTraceRefCount(0xc00010a960, 0x40)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/funcs.go:405 +0x152
google.golang.org/grpc/internal/channelz.(*channelTrace).clear(0xc0002ffe00)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:663 +0x8a
google.golang.org/grpc/internal/channelz.(*channel).deleteSelfIfReady(0xc0009f1d40)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:288 +0x1d3
google.golang.org/grpc/internal/channelz.(*channel).triggerDelete(0xae8e00?)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/types.go:229 +0x1d
google.golang.org/grpc/internal/channelz.(*channelMap).removeEntry(0xc00010a960, 0x3f)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/funcs.go:396 +0x139
google.golang.org/grpc/internal/channelz.RemoveEntry(0xc000eef770)
	/home/runner/work/grpc-go/grpc-go/internal/channelz/funcs.go:282 +0x3e
google.golang.org/grpc.(*ClientConn).Close(0xc00047f400)
	/home/runner/work/grpc-go/grpc-go/clientconn.go:1292 +0x272
```

Essentially, what's happening here is:
- Close the channel and delete the channelz ID for it
- Remove the channelz entity from the global map
- Tell the channel to delete itself and its child references
- The channel clears its trace log
- The trace log has references to the subchannel under the channel
- Decrementing the final reference to the subchannel instructs the channel to remove the subchannel from itself
- That implies that the channel should attempt to delete itself
- That means that it needs to clear out the trace log, which leads to taking the channelTrace mutex a second time

To fix this deadlock, I did what was done elsewhere in this same code: set a boolean to indicate the trace is already being cleared (`closeCalled` elsewhere).  That should fix things for now.

Medium-term, I'm pretty worried about this code.  It's not very understandable and all the back-references are problematic, because they could have similar types of races, especially if multiple things are being deleted at the same time.  I don't believe that removing a child from a parent should lead to the parent attempting to delete itself, and I also don't think removing a reference to an entity should ever make that entity delete itself from its parent, but changing these things leads to failures.  A bigger rewrite is probably called for here.

**Comment by easwars** [comment]
LGTM for the last commit.

**Pull request merged**